### DISCUSSION TITLE
Wasm support

### DISCUSSION THREAD

**Discussion by Jak2k** [ratatui/ratatui#640] [Ideas]
## Problem

My tui needs to be accessible in the browser.

## Solution

wasm

## Alternatives

Using a GUI.

**Comment by joshka** [comment]
@aschey has done a PoC for this at https://github.com/aschey/ratatui-wasm
They mentioned it in https://discord.com/channels/1070692720437383208/1072906183104528434/1156033015202451547
There are a few other mentions of WASM in the discord that might worth looking into to find out if other people have done anything

**Comment by InDieTasten** [comment]
I think having this feature would be great for being able to demo a lot of TUI projects. I would definitely love to be able to showcase my TUI apps inside xterm.js.

The examples from raratui itself could also be showcased in the browser.

**Comment by gold-silver-copper** [comment]
bevy_ratatui allows you to compile to wasm

https://github.com/gold-silver-copper/bevy_ratatui

**Reply by InDieTasten** [reply]
Interesting. Would you care to explain how it works? Is there a way to run stdin to it and have full terminal emulation with mouse support and so on?

**Reply by gold-silver-copper** [reply]
bevy is a game engine which provides input and window handling, as well as some UI primitives

bevy_ratatui works by creating a faux terminal out of the UI primitives, and then drawing to this terminal through ratatui.

Mouse suport is possible but not automatic yet. Currently you cannot use it with xterm.js , it simply makes its own render , but perhaps that will be possible in a future update when I decouple the bevy renderer from the logic loop.
You can see an example of it running in WASM here: https://gold-silver-copper.github.io/
repository: https://github.com/gold-silver-copper/gold-silver-copper.github.io/tree/main

I will be working on more examples and features coming soon !

**Comment by jrvidal** [comment]
In a bit of convergent evolution, I ended up re-creating most of @aschey's PoC for a TUI app of mine: [https://_rvidal.gitlab.io/pneo/](https://_rvidal.gitlab.io/pneo/)

It also requires a [fork of crossterm](https://gitlab.com/_rvidal/crossterm/-/tree/0.27-pneo), though the changes are quite minimal (plus a small reimplementation of [EventStream](https://docs.rs/crossterm/0.28.1/crossterm/event/struct.EventStream.html)).

**Reply by joshka** [reply]
Cool stuff!

**Reply by orhun** [reply]
Very nice, I'm curious about why a fork of crossterm is needed. Asking this question for people who might stumble upon the same challenges in the future :)

**Reply by jrvidal** [reply]
The [changes](https://gitlab.com/_rvidal/crossterm/-/compare/master...0.27-pneo?from_project_id=62172655&straight=false) pretty much speak for themselves. But in a nutshell:
* `crossterm` does not compile in `wasm32-unknown-unknown` right out of the bat, so a bunch of `unimplemented!()` placeholders are needed.
* `EventStream` is not usable as is, b/c it relies on threads and the UNIX machinery for TTYs. But part of the implementation is very much reusable: the parsing logic for data coming out of "stdin". So I simply exposed that.

**Comment by NfNitLoop** [comment]
I've taken a different approach — Instead of patching crossterm (or a different library) to work in WASM, I just wrote a new implementation of Ratatui's `Backend` trait that works in WASM.

I didn't want to go the route of adding wasm "support" to crossterm because a lot of what crossterm covers can't actually be supported in Wasm. It can't call Windows APIs to set terminal colors. It can't query the tty to inspect its size. Etc. 

In comparison, the `Backend` trait is pretty small and was (mostly) simple to implement. I just delegate to the `anes` crate to output "standard" ANSI codes in all cases. (So YMMV on older terminals. But… TBH I'm not that interested in supporting them.)

Let me know if you find this useful!

https://crates.io/crates/ratatui-wasm-backend

**Reply by orhun** [reply]
That's pretty cool! I'm curious how this compares to [`ratzilla`](https://github.com/orhun/ratzilla) and what kind of applications you can build with it  👀

**Reply by NfNitLoop** [reply]
Oh, thanks for the link. I think I'd watched a talk on ratzilla and forgotten the name.  That repo says:

> Build terminal-themed web applications with Rust and WebAssembly. Powered by [Ratatui](https://ratatui.rs/).

They offer [two backends][1]: `DomBackend` and `CanvasBackend`. So it seems they're really focused on WASM *in the browser*.

[1]: https://docs.rs/ratzilla/latest/ratzilla/backend/index.html

ratatui-wasm-backend targets the terminal. It outputs a stream of bytes, including ANSI escape sequences for styling and cursor positioning.

Now, I've seen third-party libraries for taking "terminal" byte streams and rendering them to the web. Maybe someone could pair that with `ratatui-wasm-backend`'s output to get it to work on the web?

Then you could use the same backend and have it render in the terminal or on the web depending on where it's deployed. Though, TBH, Ratatui's Backend is already pluggable, so I don't see why you wouldn't just use one of Ratzilla's backends on the web and something like `ratatui-wasm-backend` in the terminal.

Like Ratatui itself, it seems both ratzilla and I took the same approach to handling input -- We don't. 😆  In the talk I skimmed on Ratzilla, they say to just listen for DOM events for input.  For `ratatui-wasm-backend` I didn't really provide a single way to handle input. But I did [expose a feature to enable the `parser` feature of anes][2] if you want to use that.  ([Example][3] of how to use that.)

[2]: https://github.com/NfNitLoop/ratatui-wasm-backend/blob/a65e00877c4c9a3469671c25945b43f1585a8f4a/crates/ratatui-wasm-backend/Cargo.toml#L16-L24

[3]: https://github.com/NfNitLoop/ratatui-wasm-backend/blob/a65e00877c4c9a3469671c25945b43f1585a8f4a/crates/regtest/src/lib.rs#L62-L68

I'm CodyC/`@nfnitloop` in the Ratatui Discord if you/anybody want to chat more about it. 😊

**Reply by j-g00da** [reply]
Hey, this made me think if it would be possible to combine this with [WAMR](https://github.com/bytecodealliance/wasm-micro-runtime) (and [wamr-rust-sdk](https://github.com/bytecodealliance/wamr-rust-sdk)) in order to create a WASM-based alternative to [mousefood](https://github.com/j-g00da/mousefood). 👀

**Reply by NfNitLoop** [reply]
> WASM-based alternative to mousefood

I'm not sure what you mean.  I'm not familiar w/ mousefood but it says it's "embedded-graphics backend for Ratatui". I feel like a WASM runtime might not be great in the embedded space where resources are usually pretty constrained?

> if it would be possible to combine this with WAMR

I'm not familiar with WAMR, and it's not on my list of things to support right now. I'm mostly targeting JavaScript, and primarily Deno, since that's what I use. BUT, the current interface between ratatui-wasm-backend and JavaScript is pretty small.  It just needs:

 *  some callback to fetch the current size of the terminal. (Ratatui checks this before rendering each frame.) 
 * A sync function to send bytes to. (usually: stdout, but do whatever you want with the bytes.)
 
 If there's a way to make that more generic to support WAMR, I'd be happy to take a look at a PR. 😊
